src/Compilers/CSharp/Portable/Errors/DiagnosticInfoWithSymbols.cs

```diff
@@ -20,5 +24,17 @@ internal DiagnosticInfoWithSymbols(bool isWarningAsError, ErrorCode errorCode, o
         {
             this.Symbols = symbols;
         }
+
+        public override bool Equals(object obj)
+        {
+            return obj is DiagnosticInfoWithSymbols diws && 
+                   this.Symbols.SequenceEqual(diws.Symbols) &&
+                   base.Equals(obj);
```

Comment:
Consider adding a typed `Equals` method to `DiagnosticInfo`.  This would allow us to pass a strongly typed object to the base and avoid yet another cast in this code path.